class-wc-gateway-komoju.php

```diff
@@ -14,186 +14,195 @@
+    $this->title        = $this->get_option( 'title' );
+    $this->description  = $this->get_option( 'description' );
+    $this->instructions = $this->get_option( 'instructions', $this->description );
+    // Filters
+    // Actions
+    add_action( 'woocommerce_update_options_payment_gateways_' . $this->id, array( $this, 'process_admin_options' ) );
+    if ( ! $this->is_valid_for_use() ) {
+      $this->enabled = 'no';
+      WC_Gateway_Komoju::log( 'is not valid for use. No IPN set.' );
+    } else {
+      include_once( 'includes/class-wc-gateway-komoju-ipn-handler.php' );
+      new WC_Gateway_Komoju_IPN_Handler( $this->webhookSecretToken, $this->secretKey, $this->invoice_prefix );
+    }
+  }
+
+  /**
+   * Logging method
+   * @param  string $message
+   */
+  public static function log( $message ) {
+    if ( self::$log_enabled ) {
+      if ( empty( self::$log ) ) {
+        self::$log = new WC_Logger();
+      }
+      self::$log->add( 'komoju', $message );
+    }
+  }
+
+  /**
+   * Check if this gateway is enabled and available in the user's country
+   *
+   * @return bool
+   */
+  public function is_valid_for_use() {
+    return in_array( get_woocommerce_currency(), apply_filters( 'woocommerce_komoju_supported_currencies', array( 'JPY' ) ) );
+  }
+
+  /**
+   * Admin Panel Options
+   */
+  public function admin_options() {
+    if ( $this->is_valid_for_use() ) {
+      parent::admin_options();
+    } else {
+?>
+      <div class="inline error"><p><strong><?php _e( 'Gateway Disabled', 'komoju-woocommerce' ); ?></strong>: <?php _e( 'Komoju does not support your store currency.', 'komoju-woocommerce' ); ?></p></div>
+<?php
+    }
+  }
+
+  /**
+   * Initialise Gateway Settings Form Fields
+   */
+  public function init_form_fields() {
+    $this->form_fields = include( 'includes/settings-komoju.php' );
+  }
+
+  /**
+   * Process the payment and return the result
+   *
+   * @param int $order_id
+   * @return array
+   */
+  public function process_payment( $order_id ) {
+    include_once( 'includes/class-wc-gateway-komoju-request.php' );
+    $order          = wc_get_order( $order_id );
+    $komoju_request = new WC_Gateway_Komoju_Request( $this );
+    $payment_method = sanitize_text_field($_POST['komoju-method']);
+
+    return array(
+      'result'   => 'success',
+      'redirect' => $komoju_request->get_request_url( $order, $payment_method )
+    );
+  }
+
+
+  /**
+   * Payment form on checkout page
+   */
+  public function payment_fields() {
+    $this->komoju_method_form();
+  }
+
+  /**
+   * Form to choose the payment method
+   */
+  private function komoju_method_form(  $args = array(), $fields = array()  ) {
+    $default_args = array(
+      'fields_have_names' => true,
+    );
+
+    $args = wp_parse_args( $args, apply_filters( 'woocommerce_komoju_method_form_args', $default_args, $this->id ) );
+
+    $komoju_client = new KomojuApi( $this->secretKey );
+    $methods = $komoju_client->paymentMethods();
```

Comment:
If the request fails what would happen? Would it just not render the HTML for the KOMOJU payment methods, or would it prevent the user from completing the payment & order?

Reply:
it wont render the fields since `$methods` is empty - maybe we should have a fallback message here, something like 'encountered a problem communicating with KOMOJU servers' or similar

Reply:
I don't know if we (or our merchants) would want to display that to the users, in the case it fails it probably shouldn't be noticeable by a user, other than they don't have the KOMOJU options to pay with. But we probably should log something for the WooCommerce admins to be aware of.

I'm not sure how the `komoju-php` library works, if it receives a 4xx or a 5xx, will it fail silently or does it raise an exception? And if it does raise an exception then how does WordPress/WooCommerce handle it?

Reply:
I can test this out this afternoon

Reply:
I tried removing the secret key and I got this:
![image](https://user-images.githubusercontent.com/12392541/100946174-e4fa4a80-3556-11eb-86de-76e537411418.png)

So it looks like the komoju-php library will throw an error. I think it's showing the entire error because we're in development. I disabled development mode (Logging into the bash container, commenting out the `WP_DEBUG` and `WP_DEBUG_LOG` lines in /var/www/html/wp-config.php, then restarting the docker containers) and got this:

![image](https://user-images.githubusercontent.com/12392541/100946827-4ec72400-3558-11eb-893f-3bb51dcd5f33.png)

Do we want to show this to the users of the WP site?

Reply:
no, I don't think we do 😅  let me handle the exception and come up with something else

Reply:
If you need a hand let me know, happy to pair on this, or brainstorm some ideas 😄